drivers: ieee802154: fix ACK header IE related bugs

maciejbaczmanski opened the pull request:
fix bugs in current implementation of ieee802154 driver:
- after refactoring `set_vendor_ie_header_lm`, case when `link_metrics_data_len == 0` has been ignored. Fixed by refactoring the code again.
- Resetting RCP stack by openthread has been handled incorrectly (`-ENOTSUP` returned). Fixed by implementing `otPlatRadioResetCsl` in upstream OpenThread and adding `clear_ie` flag.
- refactoring of enh ACK header IE resulted with returning `-ENOTSUP` when `ack_ie.header_ie == NULL` (set when csl period set to 0 in `otPlatRadioEnableCsl`), because
`ieee802154_header_ie_get_element_id` accessed `NULL` pointer.
- `struct ieee802154_config config` is a struct containing an union. Members of `config` were accessed incorrectly in `otPlatRadioEnableCsl`. Fixed by initializing `config` with `0` and accessing one member at a time.

maciejbaczmanski commented:
DNM until https://github.com/zephyrproject-rtos/zephyr/pull/67979 is merged (tested with changes from given PR. I'll rebase this commit after merge, but opened this PR earlier to have time to resolve any comments)

jciupis commented on include/zephyr/net/ieee802154_radio.h:
```diff
@@ -1241,6 +1241,11 @@ struct ieee802154_config {
 			uint16_t short_addr;
+
+			/**
+			 * Flag for clearing enh ACK header IE (e.g during stack reset)
```
Please improve the documentation, it's unclear what effect setting this flag has on the resulting configuration. For instance, is it possible to clear all IEs and set a new one in just one call to `configure`? It doesn't follow from the documentation and requires reading the implementation.

Personally, I'd also suggest to name it `purge` instead of `clear` to strongly imply that it removes all existing entries, but that's just a preference I guess.

maciejbaczmanski replied:
Extended documentation. Please take a look if it makes sense now

jciupis commented on drivers/ieee802154/ieee802154_nrf5.c:
```diff
@@ -900,34 +900,43 @@ static int nrf5_configure(const struct device *dev,
+			nrf_802154_ack_data_clear(short_addr_le, false, NRF_802154_ACK_DATA_IE);
+			nrf_802154_ack_data_clear(ext_addr_le, true, NRF_802154_ACK_DATA_IE);
+		} else {
+			element_id = ieee802154_header_ie_get_element_id(config->ack_ie.header_ie);
+
+			if (element_id != IEEE802154_HEADER_IE_ELEMENT_ID_CSL_IE &&
+			    (!IS_ENABLED(CONFIG_NET_L2_OPENTHREAD) ||
+			     element_id != IEEE802154_HEADER_IE_ELEMENT_ID_VENDOR_SPECIFIC_IE)) {
+				return -ENOTSUP;
+			}
```
minor: I'd suggest to group all Thread-specific checks together.
```suggestion
			element_id = ieee802154_header_ie_get_element_id(config->ack_ie.header_ie);
			bool valid_vendor_specific_ie = false;
			
#if defined(CONFIG_NET_L2_OPENTHREAD)
			uint8_t vendor_oui_le[IEEE802154_OPENTHREAD_VENDOR_OUI_LEN] =
				IEEE802154_OPENTHREAD_THREAD_IE_VENDOR_OUI;

			if (element_id == IEEE802154_HEADER_IE_ELEMENT_ID_VENDOR_SPECIFIC_IE &&
			    0 == memcmp(config->ack_ie.header_ie->content.vendor_specific.vendor_oui,
				   vendor_oui_le, sizeof(vendor_oui_le))) {
				valid_vendor_specific_ie = true;
			}
#endif

			if (element_id != IEEE802154_HEADER_IE_ELEMENT_ID_CSL_IE && !valid_vendor_specific_ie)) {
				return -ENOTSUP;
			}
```

maciejbaczmanski replied:
makes sense, fixed.

jciupis commented on modules/openthread/platform/radio.c:
```diff
@@ -1273,6 +1270,8 @@ otError otPlatRadioEnableCsl(otInstance *aInstance, uint32_t aCslPeriod, otShort
 		return OT_ERROR_FAILED;
 	}
+	config.ack_ie.short_addr = aShortAddr;
+	config.ack_ie.ext_addr = aExtAddr->m8;
```
What if `aExtAddr == NULL`? AFAIK this line becomes undefined behavior then.

jciupis commented on modules/openthread/platform/radio.c:
```diff
@@ -1297,6 +1296,23 @@ otError otPlatRadioEnableCsl(otInstance *aInstance, uint32_t aCslPeriod, otShort
+	struct ieee802154_config config = { 0 };
+	int result;
+
+	config.csl_period = 0;
```
minor: This seems unnecessary if the struct is set to all zeros above anyway.

maciejbaczmanski replied:
removed

jciupis reviewed (changes requested):
Please separate commits into bisectable, atomic changes or alternatively explain why that's impractical in this case.

maciejbaczmanski committed:
drivers: ieee802154: fix ACK header IE implementation

- In `set_vendor_ie_header_lm`, case when
`link_metrics_data_len == 0` has been ignored.
This commit fixes that by setting `header_ie->length = 0`
before returning.
- current implementation of enh ACK header IE returns
`-ENOTSUP` when `ack_ie.header_ie == NULL` or
`ack_ie.header_ie->length == 0`. This commit fixes that by
refactoring checks in `nrf5_configure`.

Co-authored-by: Przemyslaw Bida <przemyslaw.bida@nordicsemi.no>

Signed-off-by: Maciej Baczmanski <maciej.baczmanski@nordicsemi.no>

maciejbaczmanski committed:
drivers: ieee802154: fix handling of `struct ieee802154_config`

`struct ieee802154_config config` is a struct containing an
union. Members of `config` were accessed incorrectly in
`otPlatRadioEnableCsl`. Fix by initializing `config` with `0`
and accessing one member at a time.

Signed-off-by: Maciej Baczmanski <maciej.baczmanski@nordicsemi.no>

maciejbaczmanski committed:
net: openthread: upmerge to `7761b81`

additionaly, implement `otPlatRadioResetCsl` functionality

Signed-off-by: Maciej Baczmanski <maciej.baczmanski@nordicsemi.no>

maciejbaczmanski commented:
> Please separate commits into bisectable, atomic changes or alternatively explain why that's impractical in this case.

I've split the commit into 3. Decided not to split:
```
drivers: ieee802154: fix ACK header IE implementation 

- In `set_vendor_ie_header_lm`, case when
`link_metrics_data_len == 0` has been ignored.
This commit fixes that by setting `header_ie->length = 0`
before returning.
- current implementation of enh ACK header IE returns
`-ENOTSUP` when `ack_ie.header_ie == NULL` or
`ack_ie.header_ie->length == 0`. This commit fixes that by
refactoring checks in `nrf5_configure`.
```

as refactoring of `set_vendor_ie_header_lm` and `nrf5_configure` are strictly depending (fixing one of them before another would still result in an incorrect implementation).

jciupis reviewed (dismissed):
LGTM, thanks!
